components/brave_wallet/browser/eth_pending_tx_tracker.cc

```diff
@@ -0,0 +1,151 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/eth_pending_tx_tracker.h"
+
+#include <utility>
+
+#include "base/logging.h"
+#include "base/synchronization/lock.h"
+#include "brave/components/brave_wallet/browser/eth_json_rpc_controller.h"
+#include "brave/components/brave_wallet/browser/eth_nonce_tracker.h"
+
+namespace brave_wallet {
+
+EthPendingTxTracker::EthPendingTxTracker(EthTxStateManager* tx_state_manager,
+                                         EthJsonRpcController* rpc_controller,
+                                         EthNonceTracker* nonce_tracker)
+    : tx_state_manager_(tx_state_manager),
+      rpc_controller_(rpc_controller),
+      nonce_tracker_(nonce_tracker),
+      weak_factory_(this) {}
+EthPendingTxTracker::~EthPendingTxTracker() = default;
+
+void EthPendingTxTracker::UpdatePendingTransactions() {
+  base::Lock* nonce_lock = nonce_tracker_->GetLock();
+  if (!nonce_lock->Try())
+    return;
+
+  auto pending_transactions = tx_state_manager_->GetTransactionsByStatus(
+      EthTxStateManager::TransactionStatus::SUBMITTED,
+      base::Optional<EthAddress>());
+  for (const auto& pending_transaction : pending_transactions) {
+    if (IsNonceTaken(pending_transaction)) {
+      DropTransaction(pending_transaction);
+      nonce_lock->Release();
+      return;
+    }
+    std::string id = pending_transaction.id;
+    rpc_controller_->GetTransactionReceipt(
+        pending_transaction.tx_hash,
+        base::BindOnce(&EthPendingTxTracker::OnGetTxReceipt,
+                       weak_factory_.GetWeakPtr(), std::move(id)));
+  }
+
+  nonce_lock->Release();
+}
+
+void EthPendingTxTracker::ResubmitPendingTransactions() {
```

Comment:
Just noting we probably want to be able to resubmit a single particular pending transaction too

Reply:
I think that would be SpeedupTransaction which uses same nonce but with 10% higher gas price